### ISSUE TITLE
Upon installing zsh I get this error: _zsh_autosuggest_highlight_reset:3: maximum nested function level reached; increase FUNCNEST?

### ISSUE THREAD

**Issue by JimPix1** [zimfw/zimfw#449]
On a fresh install of zsh and zim I get this error
_zsh_autosuggest_highlight_reset:3: maximum nested function level reached; increase FUNCNEST?
I have done nothing to the configs whatsoever, and I have also tried reinstalling
I am using EndeavourOS

**Comment by ericbn** [comment]
Maybe you're sourcing the plugin more than once? Once with Zim and once directly in your `.zshrc` or with a second plugin manager? Check https://github.com/zsh-users/zsh-autosuggestions/issues/187#issuecomment-240310586 and https://github.com/zsh-users/zsh-autosuggestions/issues/84#issuecomment-173089220

**Comment by JimPix1** [comment]
I'm not using another plugin manager and I did an empty .zshrc to test it as well

**Comment by JimPix1** [comment]
It doesn't make any sense to me

**Comment by romkatv** [comment]
It makes sense if you are using Kitty. Are you?

**Comment by JimPix1** [comment]
I've been testing this in kitty, yes

**Comment by romkatv** [comment]
The new version of kitty (0.24.0) injects zsh code into zshenv (before your own zshenv) and that code is buggy. This particular bug has been fixed in master, so if you can build from source, that'll help. Alternatively, add this line to `~/.zshenv`: `unset KITTY_SHELL_INTEGRATION`. Another alternative is to disable shell integration in `kitty.conf`.

**Comment by JimPix1** [comment]
Oh, thank you very much!

**Comment by romkatv** [comment]
The fact that terminals (well, one terminal) now inject code into your shell is really quite awful. From the point of view of zsh plugin and framework maintainers it's terrifying.

**Comment by JimPix1** [comment]
Yeah, that really shouldn't be a thing and I hope it's removed entirely from kitty since aside from well, this, it's a very good terminal
Thanks again for fixing this!

**Comment by danielshahaf** [comment]
A bit more context.

Kitty shell integration docs: https://github.com/kovidgoyal/kitty/blob/de3b9634ac15a81f8e99fa7410456792b91401be/docs/shell-integration.rst

Kitty bug report requesting to not modify bashrc by default: https://github.com/kovidgoyal/kitty/issues/4429.  Closed as WONTFIX.

**Comment by romkatv** [comment]
Zsh integration in kitty does not modify rc files. Before starting zsh it points `ZDOTDIR` to its own directory with `.zshenv` in it. This `.zshenv` installs hooks and restores the original value of `ZDOTDIR`. This ingenious solutions makes finding the culprit extremely difficult. The usual tools such as `zsh --sourcetrace` or `zsh -x` won't help because kitty injects code only into the top-level shell. Even `exec zsh` rids shell of kitty integration.

- Terminal code: https://github.com/kovidgoyal/kitty/blob/57888af4ea321c545612b33a2fc76b3ef0b93444/kitty/shell_integration.py#L70-L97
- Shell code: https://github.com/kovidgoyal/kitty/tree/872cd7b2d5f696229755d877e3f36bdb386490c7/shell-integration/zsh

I wrote both of these pieces. To be more precise, I rewrote them. My preference would be for kitty to not inject any code into zsh but this is not something I can do. As the next best alternative I've tried to make the injected code less likely to break stuff. I've also added a way to opt out from kitty integration from within zsh. I use it in my projects.

The only kitty-specific piece of code in kitty's integration is `_ksi_debug_print` function, which is meant for debugging. Other than that it's your regular terminal-agnostic zsh plugin stuff: define completion functions, change cursor shape based on the current keymap, update terminal title on precmd and preexec, etc. All of these already exist as standalone plugins.

Most of the discussion around kitty integration has happened here: https://github.com/kovidgoyal/kitty/issues/3948

**Comment by danielshahaf** [comment]
@romkatv Thanks a lot for all the extra information.  To summarize:

- Kitty's magic only affect the zsh instance started by kitty.  It doesn't affect any other zsh process started within a kitty-owned pty, even those started via `exec zsh`.
- `zsh -f` will skip kitty's integrations.  Emptying ~/.zshenv, ~/.zprofile, ~/.zshrc, ~/.zlogin will not.

One more question, if our zimfw hosts will permit it: Suppose someone reports a bug to zsh upstream or to a random zsh plugin, what can the maintainers ask the reporter to execute in a running zsh instance in order to determine whether the reporter has kitty integration enabled in that zsh instance?  The BOFH answer is `ps -p $PPID | grep -i kitty`, but…

**Comment by romkatv** [comment]
> * Kitty's magic only affect the zsh instance started by kitty.  It doesn't affect any other zsh process started within a kitty-owned pty, even those started via `exec zsh`.

Correct as far as automatic shell integration via `ZDOTDIR` override is concerned. One can also enable kitty shell integration by placing the following stanza in `.zshrc`:

```zsh
if [[ -n $KITTY_INSTALLATION_DIR ]]; then
  export KITTY_SHELL_INTEGRATION="enabled"
  autoload -Uz -- "$KITTY_INSTALLATION_DIR"/shell-integration/zsh/kitty-integration
  kitty-integration
  unfunction kitty-integration
fi
```

Or this:

```zsh
if [[ -n $KITTY_INSTALLATION_DIR ]]; then
  source -- "$KITTY_INSTALLATION_DIR"/shell-integration/zsh/kitty.zsh
fi
```

This is a more conventional approach.

> * `zsh -f` will skip kitty's integrations.  Emptying ~/.zshenv, ~/.zprofile, ~/.zshrc, ~/.zlogin will not.

Right. If you instruct kitty to start `zsh -f` instead of the user's login shell, automatic shell integration won't work. However, if one can set `shell zsh -f` in `kitty.conf`, they can also set `shell_integration disabled` in there.

> One more question, if our zimfw hosts will permit it: Suppose someone reports a bug to zsh upstream or to a random zsh plugin, what can the maintainers ask the reporter to execute in a running zsh instance in order to determine whether the reporter has kitty integration enabled in that zsh instance? The BOFH answer is `ps -p $PPID | grep -i kitty`, but…

Kitty shell integration defines parameters and functions with the prefix `_ksi_`. The presence of `_ksi_state` parameter indicates that integration is enabled.

Integration can be disabled by placing `unset KITTY_SHELL_INTEGRATION` in the user's `.zshenv`, so when in doubt you can ask the reporter to do that and see whether it fixes the problem.

**Comment by smdotdev** [comment]
> The new version of kitty (0.24.0) injects zsh code into zshenv (before your own zshenv) and that code is buggy. This particular bug has been fixed in master, so if you can build from source, that'll help. Alternatively, add this line to `~/.zshenv`: `unset KITTY_SHELL_INTEGRATION`. Another alternative is to disable shell integration in `kitty.conf`.

Thank you so much
